### PR TITLE
Don't start a server if there's unbound ports (for v1.4.x branch)

### PR DESCRIPTION
**Pull request by jtattermusch**
Tentative fix for #10918

### CODE REVIEW

**Comment by jtattermusch** [comment]
CC @chwarr

File: src/csharp/Grpc.Core.Tests/ServerTest.cs [hunk 1]
```diff
@@ -81,6 +82,21 @@ public void PickUnusedPort()
         }
 
         [Test]
+        public void StartThrowsWithUnboundPorts()
+        {
+            int twiceBoundPort = 9999;
```
Comment (chwarr) [inline src/csharp/Grpc.Core.Tests/ServerTest.cs]: I don't know this code base's style well, but `twiceBoundPort` can be marked `const`.

File: src/csharp/Grpc.Core/Server.cs [hunk 2]
```diff
@@ -316,6 +319,20 @@ private void AllowOneRpc(CompletionQueueSafeHandle cq)
...
+        /// <summary>
+        /// Checks that all ports have been bound successfully.
+        /// </summary>
+        private void CheckPortsBoundSuccessfully()
+        {
+            lock (myLock)
+            {
+                if (!ports.All((port) => port.BoundPort != 0))
+                {
+                    throw new IOException("Failed to bind some of ports exposed by the server.");
```
Comment (chwarr) [inline src/csharp/Grpc.Core/Server.cs]: If the unbound ports (or at least one of them) were included in the exception message, that would make it more actionable. Though, I'm not sure whether having custom messages in exceptions is OK in this code base. (Some people like to keep all their exception messages static to help with localization, for example.)

Something like this?

```
var unboundPort = ports.FirstOrDefault(port => port.BoundPort == 0);
if (unboundPort != null)
{
    // may need to pass a culture: don't know this code base's style
    throw new IOException(
        string.Format("Failed to bind port \"{0}:{1}\"", unboundPort.Host, unboundPort.Port));
}
```

Comment (jtattermusch) [inline src/csharp/Grpc.Core/Server.cs]: Good suggestion, I used your snippet.

**Pull request merged**